chapter_frame.go

```diff
@@ -0,0 +1,149 @@
+package id3v2
+
+import (
+	"encoding/binary"
+	"io"
+	"time"
+)
+
+const (
+	nanosInMillis = 1000000
+	IgnoredOffset = 0xFFFFFFFF
+)
+
+// ChapterFrame is used to work with CHAP frames
+// according to spec from http://id3.org/id3v2-chapters-1.0
+// This implementation only supports single TIT2 subframe (Title field).
+// All other subframes are ignored.
+// If StartOffset or EndOffset == id3v2.IgnoredOffset, then it should be ignored
+// and StartTime or EndTime should be utilized
+type ChapterFrame struct {
+	ElementID   string
+	StartTime   time.Duration
+	EndTime     time.Duration
+	StartOffset uint32
+	EndOffset   uint32
+	Title       *TextFrame
+	SubTitle    *TextFrame
+}
+
+func (cf ChapterFrame) Size() int {
+	size := encodedSize(cf.ElementID, EncodingISO) +
+		1 + // trailing zero after ElementID
+		4 + 4 + 4 + 4 // (Start, End) (Time, Offset)
+	if cf.Title != nil {
+		size = size +
+			frameHeaderSize + // Title frame header size
+			cf.Title.Size()
+	}
+	if cf.SubTitle != nil {
+		size = size +
+			frameHeaderSize + // SubTitle frame header size
+			cf.SubTitle.Size()
+	}
+	return size
+}
+
+func (cf ChapterFrame) UniqueIdentifier() string {
+	return cf.ElementID
+}
+
+func (cf ChapterFrame) WriteTo(w io.Writer) (n int64, err error) {
+	return useBufWriter(w, func(bw *bufWriter) {
+		bw.EncodeAndWriteText(cf.ElementID, EncodingISO)
+		bw.WriteByte(0)
+		// nanoseconds => milliseconds
+		binary.Write(bw, binary.BigEndian, int32(cf.StartTime/nanosInMillis))
+		binary.Write(bw, binary.BigEndian, int32(cf.EndTime/nanosInMillis))
+
+		binary.Write(bw, binary.BigEndian, cf.StartOffset)
+		binary.Write(bw, binary.BigEndian, cf.EndOffset)
+
+		if cf.Title != nil {
+			writeFrame(bw, "TIT2", *cf.Title, true)
+		}
+
+		if cf.SubTitle != nil {
+			writeFrame(bw, "TIT3", cf.SubTitle, true)
+		}
+	})
+}
+
+func parseChapterFrame(br *bufReader) (Framer, error) {
+	ElementID := br.ReadText(EncodingISO)
+	var startTime uint32
+	var startOffset uint32
+	var endTime uint32
+	var endOffset uint32
+
+	if err := binary.Read(br, binary.BigEndian, &startTime); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(br, binary.BigEndian, &endTime); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(br, binary.BigEndian, &startOffset); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(br, binary.BigEndian, &endOffset); err != nil {
+		return nil, err
+	}
+
+	var title TextFrame
+	var subTitle TextFrame
+
+	// borrowed from parse.go
+	buf := getByteSlice(32 * 1024)
+	defer putByteSlice(buf)
+	for {
+		// no way to determine whether this should be true or not
```

Comment:
synchSafe is true when tag version is 4. You have to add an argument `synchSafe bool` all parsers functions and provide `tag.Version() === 4` to them [like here](https://github.com/bogem/id3v2/blob/master/parse.go#L67)